### ISSUE TITLE
Pod with volume attached to a node can't failover when the node goes down

### ISSUE THREAD

**Issue by cvvz** [kubernetes/kubernetes#112016]
### What happened?

1. Create a deployment with N replicas and AzureDisk Volume (for example).
2. Shutdown the Node in which Pods resident.
3. Wait until the Node goes into `NotReady` state and Pods go into `Terminating` state( it can't be really deleted since Node is down ) and new Pods are generated automatically.
4. These new Pods stay in `containerCreating` state and will never go into `Ready` state since volume can not be detached from the failed node.

### What did you expect to happen?

Volume can be detached successfully after the node goes down, and new Pods can be created successfully on another node, which means Pod can failover successfully.

### How can we reproduce it (as minimally and precisely as possible)?

As described above.

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by niulechuan** [comment]
``Detach`` should be called only after a volume was successfully unmounted by all nodes' Kubelet. 

In Kubelet a volume was unmounted only after the pod that uses the volume was deleted from Apiserver and made sure all container was stopped. Therefore A pod that existed in Apiserver which means unmount can't able to call.

So we cannot just call ``Detach`` in these circumstances.

**Comment by cvvz** [comment]
@niulechuan Thanks for your explanation! Seems like I consider too superficially. 

Do you think this situation is a problem that worth to solve? If so, any good idea to solve it?

**Comment by cvvz** [comment]
`azuredisk-csi-driver v2` may provide a solution to work around this situation:
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/docs/design-v2.md

**Comment by cvvz** [comment]
This issue is noted in: https://github.com/kubernetes-sigs/azuredisk-csi-driver/tree/master/docs/known-issues/node-shutdown-recovery. It is a common issue.

**Comment by cvvz** [comment]
> azuredisk-csi-driver v2 may provide a solution to work around this situation.

Actually I've tested with azuredisk-csi-driver v2 today but seems like this scenario is not resolved, it act just like V1. The newly created Pod was in `ContainerCreating` state forever because of multi-attach. We need to detach the disk from the former attached node manualy to work around.

**Comment by andyzhangx** [comment]
https://github.com/kubernetes-sigs/cloud-provider-azure/pull/4008 would fix the issue on Azure

**Comment by xing-yang** [comment]
@cvvz @andyzhangx Can this issue be closed since the non-graceful node shutdown feature should have resolved this?

**Comment by mc2285** [comment]
I kindly request to reopen.

Non-graceful node shutdown does not address this problem. Non-graceful node shutdown aims to address a completely separate issue: if I know a node went down in an unexpected way, how do I manualy tell `k8s` that I know the node is down and resources like `VolumeAttachments` and `Pods` can be force-deleted safely because I am sure they are not active on the dead node.

The problem at least I see here regards what should happen automatically. As of version `v1.32.2` after the 6-minute timer for forced detach expires, contrary to expectations, nothing happens as forced detach will only occur if the related pod is no longer running. This however will not be the case if a node goes down unexpectedly as the pod will be then stuck terminating. 

What I, and from my knowledge many other users, would expect to happen is for pods on nodes that have been `NotReady` for some time to be force-terminated automatically to allow rescheduling. Obviously many workloads cannot allow such a behavior as has been explained a countless number of times in this bug tracker. However, many also require it and it would be nice to see an upstream option to enable such a behavior, given that most of the required code is already in place (it would basically be automatic non-graceful node shutdown after a period of time).

Judging from the fact that a large commercial CSI provider like Longhorn backed by an enormous company like SUSE has decided to implement (such a feature)[https://longhorn.io/docs/1.9.1/high-availability/node-failure/] and has been supporting it for years, there is most definitely a use case for such a behavior.

I know this looks like a perfect use case for an operator, however, to my knowledge, as of today, there is no widely-adopted one which leads to many vendors developing in-house solutions for this problem. Again, it would be nice to see an upstream solution to such a common problem as it has been requested a countless number of times. Many users simply need the option to force-terminate pods on nodes that are down for a longer period of time.

The argument that it could lead to data loss no longer holds with non-graceful node shutdown in place. Non-graceful node shutdown also could but we have the option to tell `k8s` that we know what we are doing and the node is surely down. It would only be logical to also allow users to tell `k8s` that we know what we are doing and please force-terminate pods on nodes that have been down for a specified amount of time. It would be nice if `kubelet` did the same on its side if `apiserver` cannot be contacted for a specified interval and those two optional features would nicely complement each other as not all workloads are cloud-native and clustered. In fact, most aren't and even some of those that are could take advantage of not having to manually clean up `StatefulSet` pods if a node goes down in an unexpected way.

If this should be moved into a separate issue let me know

**Comment by mc2285** [comment]
Related: https://github.com/kubernetes/kubernetes/issues/116618
